text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,1042 @@
+for types whose `Default` impl's `default` method satisfies all rules that `const fn` must satisfy
+(including some annotation that guarantees this won't break by accident).
+It must always be possible to call `default` outside of const contexts with no limitations on the generic parameters that may be passed.
+
+Similarly it should be possible to write `compile_time_default` in a way that also requires calls
+outside of const contexts to only pass generic parameters whose `Default::default` method satisifies
+the usual `const fn` rules. This is necessary in order to allow a const block
+(which can access generic parameters) in the function body to invoke methods on the generic parameter.
+
+So, we need some annotation that differentiates a `T: Default` bound from one that gives us the guarantees we're looking for.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Nomenclature and new syntax concepts
+
+### Const trait methods
+
+Traits can declare methods as `const`. Doing so is a breaking change, as all impls are now required to provide a `const` method,
+which existing impls can't.
+
+```rust
+trait Trait {
+    const fn method();
+}
+```
+
+These methods need to be implemented as `const`:
+
+```rust
+impl Trait for Type {
+    const fn method() {}
+}
+```
+
+### Const trait bounds
+
+Any item that can have trait bounds can also have `const Trait` bounds.
+
+Examples:
+
+* `T: const Trait`, requiring any type that `T` is instantiated with to have a trait impl with `const` methods for `Trait`.
+* `dyn const Trait`, requiring any type that is unsized to this dyn trait to have a trait impl with `const` methods for `Trait`.
+    * These are not part of this RFC because they require `const` function pointers. See [the Future Possibilities section](#future-possibilities).
+* `impl const Trait` (in all positions).
+    * These are not part of this RFC.
+* `trait Foo: const Bar {}`, requiring every type that has an impl for `Foo` (even a non-const one), to also have a trait impl with `const` methods for `Bar`.
+* `trait Foo { type Bar: const Trait; }`, requiring all the impls to provide a type for `Bar` that has a trait impl with `const` methods for `Trait`
+
+Such an impl allows you to use the type that is bound within a const block or any other const context, because we know that the type has a trait impl with `const` methods and thus
+must be executable at compile time. The following function will invoke the `Default` impl of a type at compile time and store the result in a constant. Then it returns that constant instead of computing the value every time.
+
+```rust
+fn compile_time_default<T: const Default>() -> T {
+    const { T::default() }
+}
+```
+
+### Conditionally const traits methods
+
+Traits need to opt-in to allowing their impls to have const methods. Thus you need to mark the trait as `[const]` and all the methods will become const callable.
```

Comment:
It took me a while to understand _why_ this opt-in is needed, so here is a rewording suggestion that will make it more obvious:

```suggestion
Making a trait's methods callable in const contexts imposes extra constraints on their default implementations (if any).
Therefore, traits need to opt-in to allowing their impls to have const methods.
You need to mark the trait as `[const]` and all the methods will become const callable.
```